File: androidApp/src/main/kotlin/com/ryntra/mobile/auth/OAuthCoordinator.kt
package com.ryntra.mobile.auth

import android.content.Context
import android.net.Uri
import androidx.core.content.edit
import androidx.core.net.toUri
import java.security.MessageDigest
import java.security.SecureRandom

class OAuthCoordinator(context: Context) {
    private val preferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    private val secureRandom = SecureRandom()

    fun createAuthorizationUri(): Uri {
        val state = ByteArray(32).also(secureRandom::nextBytes).toHex()
        preferences.edit { putString(STATE_KEY, state) }
        return OAUTH_START_URL.toUri()
            .buildUpon()
            .appendQueryParameter("state", state)
            .build()
    }

    fun consumeCallback(uri: Uri): OAuthCallbackResult {
        if (uri.scheme !in CALLBACK_SCHEMES || uri.host != CALLBACK_HOST || uri.path != CALLBACK_PATH) {
            return OAuthCallbackResult.Ignored
        }

        val expectedState = preferences.getString(STATE_KEY, null)
        preferences.edit { remove(STATE_KEY) }
        val returnedState = uri.getQueryParameter("state")
        if (!statesMatch(expectedState, returnedState)) {
            return OAuthCallbackResult.Failure("Sign-in failed because the OAuth state did not match.")
        }

        if (!uri.getQueryParameter("error").isNullOrBlank()) {
            return OAuthCallbackResult.Failure("Modrinth sign-in was cancelled.")
        }

        val token = uri.getQueryParameter("token")?.trim()
        if (token.isNullOrEmpty() || token.length > MAX_TOKEN_LENGTH) {
            return OAuthCallbackResult.Failure("OAuth did not return a valid access token.")
        }
        return OAuthCallbackResult.Success(token)
    }

    fun clear() {
        preferences.edit { remove(STATE_KEY) }
    }

    private fun statesMatch(expected: String?, returned: String?): Boolean {
        if (expected == null || returned == null) return false
        return MessageDigest.isEqual(
            expected.toByteArray(Charsets.UTF_8),
            returned.toByteArray(Charsets.UTF_8),
        )
    }

    private fun ByteArray.toHex(): String = joinToString(separator = "") { byte ->
        byte.toUByte().toString(radix = 16).padStart(length = 2, padChar = '0')
    }

    private companion object {
        const val PREFERENCES_NAME = "oauth_session"
        const val STATE_KEY = "expected_state"
        // The auth deployment still uses its original hostname and callback scheme.
        const val OAUTH_START_URL = "https://rinthy-auth.vercel.app/api/modrinth/start"
        val CALLBACK_SCHEMES = setOf("ryntra", "rinthy")
        const val CALLBACK_HOST = "auth"
        const val CALLBACK_PATH = "/callback"
        const val MAX_TOKEN_LENGTH = 4_096
    }
}

sealed interface OAuthCallbackResult {
    data object Ignored : OAuthCallbackResult

    data class Success(val token: String) : OAuthCallbackResult

    data class Failure(val message: String) : OAuthCallbackResult
}

File: iosApp/Ryntra/OAuthCoordinator.swift
import Foundation
import Security

final class OAuthCoordinator {
    enum CallbackResult {
        case ignored
        case success(String)
        case failure(String)
    }

    private let defaults = UserDefaults.standard
    private let stateKey = "modrinth_oauth_state"

    func createAuthorizationURL() -> URL? {
        let state = randomState()
        defaults.set(state, forKey: stateKey)

        // The auth deployment still uses its original hostname and callback scheme.
        var components = URLComponents(string: "https://rinthy-auth.vercel.app/api/modrinth/start")
        components?.queryItems = [URLQueryItem(name: "state", value: state)]
        return components?.url
    }

    func consumeCallback(_ url: URL) -> CallbackResult {
        guard let scheme = url.scheme,
              Self.callbackSchemes.contains(scheme),
              url.host == "auth",
              url.path == "/callback" else {
            return .ignored
        }

        let expectedState = defaults.string(forKey: stateKey)
        defaults.removeObject(forKey: stateKey)
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        let returnedState = components?.value(for: "state")
        guard statesMatch(expectedState, returnedState) else {
            return .failure("Sign-in failed because the OAuth state did not match.")
        }

        if let error = components?.value(for: "error"), !error.isEmpty {
            return .failure("Modrinth sign-in was cancelled.")
        }

        guard let token = components?.value(for: "token")?.trimmingCharacters(in: .whitespacesAndNewlines),
              !token.isEmpty,
              token.count <= 4_096 else {
            return .failure("OAuth did not return a valid access token.")
        }
        return .success(token)
    }

    func clear() {
        defaults.removeObject(forKey: stateKey)
    }

    private func randomState() -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            return UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
        }
        return bytes.map { String(format: "%02x", $0) }.joined()
    }

    private func statesMatch(_ expected: String?, _ returned: String?) -> Bool {
        guard let expected, let returned else { return false }
        let left = Array(expected.utf8)
        let right = Array(returned.utf8)
        guard left.count == right.count else { return false }
        return zip(left, right).reduce(UInt8(0)) { difference, pair in
            difference | (pair.0 ^ pair.1)
        } == 0
    }

    private static let callbackSchemes: Set<String> = ["ryntra", "rinthy"]
}

private extension URLComponents {
    func value(for name: String) -> String? {
        queryItems?.first(where: { $0.name == name })?.value
    }
}
